app/helpers/filter_helper.rb

```diff
@@ -0,0 +1,86 @@
+module FilterHelper
+  def build_condition_query(model_filters, query_hash, current_index)
+    current_filter = model_filters[query_hash['attribute_key']]
+
+    # Throw InvalidOperator Error if the attribute is a standard attribute
+    # and the operator is not allowed in the config
+    if current_filter.present? && current_filter['filter_operators'].exclude?(query_hash[:filter_operator])
+      raise CustomExceptions::CustomFilter::InvalidOperator.new(
+        attribute_name: query_hash['attribute_key'],
+        allowed_keys: current_filter['filter_operators']
+      )
+    end
+
+    # Every other filter expects a value to be present
+    if %w[is_present is_not_present].exclude?(query_hash[:filter_operator]) && query_hash['values'].blank?
+      raise CustomExceptions::CustomFilter::InvalidValue.new(attribute_name: query_hash['attribute_key'])
+    end
+
+    condition_query = build_condition_query_string(current_filter, query_hash, current_index)
+    # The query becomes empty only when it doesn't match to any supported
+    # standard attribute or custom attribute defined in the account.
+    if condition_query.empty?
+      raise CustomExceptions::CustomFilter::InvalidAttribute.new(key: query_hash['attribute_key'],
+                                                                 allowed_keys: model_filters.keys)
+    end
+
+    condition_query
+  end
+
+  def build_condition_query_string(current_filter, query_hash, current_index)
+    filter_operator_value = filter_operation(query_hash, current_index)
+
+    return handle_nil_filter(query_hash, current_index) if current_filter.nil?
+
+    case current_filter['attribute_type']
+    when 'additional_attributes'
+      handle_additional_attributes(query_hash, filter_operator_value)
+    else
+      handle_standard_attributes(current_filter, query_hash, current_index, filter_operator_value)
+    end
+  end
+
+  def handle_nil_filter(query_hash, current_index)
+    attribute_type = "#{filter_config[:entity].downcase}_attribute"
+    custom_attribute_query(query_hash, attribute_type, current_index)
+  end
+
+  # TODO: Change the reliance on entity instead introduce datatype text_case_insensive
+  # Then we can remove the condition for Contact
+  def handle_additional_attributes(query_hash, filter_operator_value)
+    if filter_config[:entity] == 'Contact'
+      "LOWER(#{filter_config[:table_name]}.additional_attributes ->> '#{query_hash[:attribute_key]}') " \
+        "#{filter_operator_value} #{query_hash[:query_operator]}"
+    else
+      "#{filter_config[:table_name]}.additional_attributes ->> '#{query_hash[:attribute_key]}' " \
+        "#{filter_operator_value} #{query_hash[:query_operator]} "
+    end
+  end
```

Comment:
We can abstract the query building logic for the `handle_additonal_attributes` and `default_filter` 

```rb
def handle_additional_attributes(query_hash, filter_operator_value)
  attribute_value = "#{filter_config[:table_name]}.additional_attributes ->> '#{query_hash[:attribute_key]}'"
  build_query_condition(attribute_value, filter_operator_value, query_hash[:query_operator])
end

def default_filter(query_hash, filter_operator_value)
  attribute_value = "#{filter_config[:table_name]}.#{query_hash[:attribute_key]}"
  build_query_condition(attribute_value, filter_operator_value, query_hash[:query_operator])
end

def build_query_condition(attribute_value, filter_operator_value, query_operator)
  if filter_config[:entity] == 'Contact'
    "LOWER(#{attribute_value}) #{filter_operator_value} #{query_operator}"
  else
    "#{attribute_value} #{filter_operator_value} #{query_operator}"
  end
end
```

Reply:
Have some work done on this in : https://github.com/chatwoot/chatwoot/pull/9048/files 

We will do further refactoring after both these PRs are merged in